### PR TITLE
WIP: Shaped single-dimension coordinates

### PR DESCRIPTION
**Pull request by jmilloy**
The following coordinates are generalized to handle shaped (nd-array) coordinates:
 - `ArrayCoordinates1d`
 - `StackedCoordinates`

`DependentCoordinates` are no longer needed because `StackedCoordinates` now handles that case. Therefore, `RotatedCoordinates` now subclasses `StackedCoordinates`.

### Features

 * `BaseCoordinates.ndim` property
 * `BaseCoordinates.flatten` method
 * `BaseCoordinates.reshape` method

### Bugfixes and Maintenance

 * `Coordinates.coords` property is renamed `xcoords`. It is meant as the `coords` argument to `xr.DataArray`, and was being mis-used. Use `c['lat'].coordinates` instead of `c.coords['lat']`.

### TODO

 - [x] replace ArraycoordinatesNd with generalized ArrayCoordinates1d
 - [x] replace DependentCoordinates with generalized StackedCoordinates
 - [x] update RotatedCoordinates
 - [ ] update docs and notebooks

### CODE REVIEW

**Comment by mpu-creare** [comment]
Glad you're working on this. Let me know when you want me to review.

**Comment by jmilloy** [comment]
@mpu-creare This is close to complete. the DependentCoordinates are fully removed and a lot of tests updated. I don't think have fully handled interpolating `StackedCoordinates` with ndim>1, which arise when unstacked coordinates get transformed to another crs. Basically, I think we need to `flatten`, interpolate, and then `reshape`. I wonder if you want to merge this into your interpolation branch (or the other way around)?

I also need to update the RotatedCoordinates still, and that won't take long. I'm doing that now.

**Comment by mpu-creare** [comment]
@jmilloy I'd like to merge THIS into the interpolation branch. So this will merge first. 
Thanks for the effort. 

The interpolation branch is coming along, but I'll have to split it into another PR. The Nearest Neighbor interpolator needs it's own attentions to solve some of the issues... and that's big enough for its own PR. Right now I'm just working through failed unit tests (that used to be working).

**Comment by jmilloy** [comment]
> Right now I'm just working through failed unit tests (that used to be working).

I reworked the datasource tests. I hope we don't get into anything too hairy.

**Comment by jmilloy** [comment]
@mpu-creare Okay, this is ready to merge now. **NOTE: I changed the base to the interpolation branch.** I still need to update docs and notebooks, but that's not going to stop you from merging.

File: podpac/core/coordinates/test/test_polar_coordinates.py
Comment (mpu-creare) [inline podpac/core/coordinates/test/test_polar_coordinates.py:261]: Why  not just delete the file?

File: podpac/core/coordinates/uniform_coordinates1d.py
Comment (mpu-creare) [inline podpac/core/coordinates/uniform_coordinates1d.py:274]: I  can't even imagine what a multi-dimensional uniform coordinates would look like... you could probably remove the TODO.

File: podpac/core/data/file_source.py
Comment (mpu-creare) [inline podpac/core/data/file_source.py:53]: Nice catch, thanks!

File: podpac/core/data/test/test_datasource.py
Comment (mpu-creare) [inline podpac/core/data/test/test_datasource.py:25]: Thanks for actually fixing the tests.

File: podpac/core/interpolation/interpolators.py
Comment (mpu-creare) [inline podpac/core/interpolation/interpolators.py:371]: I wonder how `git` will handle the fact that I deleted this file in the other branch... should be interesting merge. Make a note that this change is retained after the merge.

**Review by mpu-creare — APPROVED** [review]
Look great. Lots of work here. Nicely done. Merge will be tricky. Please respond to a few comments then we can merge. I"ll resolve the comments that are just comments.

File: podpac/core/coordinates/test/test_polar_coordinates.py
Comment (jmilloy) [reply podpac/core/coordinates/test/test_polar_coordinates.py:261]: Ah I overlooked polar coordinates. I should actually update the polar coordinates and then reinstate the tests.

File: podpac/core/compositor/compositor.py
Comment (jmilloy) [reply podpac/core/compositor/compositor.py:146]: Yep, that's why!

File: podpac/core/coordinates/array_coordinates1d.py
Comment (jmilloy) [reply podpac/core/coordinates/array_coordinates1d.py:74]: Yes, I guess so. Two comments:
 - request coordinates will often be ndim>1 (when we need a crs transform), but the source coordinates (which you will be selecting) will almost always be ndim=1.
 - I think we will "flatten, interpolate, reshape". But maybe that will look more like "flatten, select, interpolate, reshape".

In other words, hopefully we can just do all of the selecting on flat coordinates.

File: podpac/core/coordinates/array_coordinates1d.py
Comment (jmilloy) [reply podpac/core/coordinates/array_coordinates1d.py:405]: I already had the `ArrayCoordinatesNd` class, which subclossed `ArrayCoordinates1d' and provided support for ndim>1, and was quite short!

File: podpac/core/coordinates/coordinates1d.py
Comment (jmilloy) [reply podpac/core/coordinates/coordinates1d.py:71]: Ah, I'm actually going to fix this. The `ArrayCoordinates1d` and `UniformCoordinates1d` both implement their own `__eq__` method, which calls this one and then makes additional checks. So it is misleading here. These are just the short-cuts in order to avoid comparing coordinates unless necessary.

File: podpac/core/coordinates/rotated_coordinates.py
Comment (jmilloy) [reply podpac/core/coordinates/rotated_coordinates.py:207]: This was tricky.

The `StackedCoordinates` is fundamentally a wrapper around its `_coords` attribute, a list of `Coordinates1d`. But the `RotatedCoordinates` doesn't have one because it is parameterized instead. So this really actually is converting from `RotatedCoordinates` to `StackedCoordinates` and then calling `__getitem__` in order to get access to the `_coords` list.

I *could* split `StackedCoordinates` up into `StackedCoordinatesBase`, `StackedCoordinates`, `RotatedCoordinates` (sort of analogous to `Coordinates1d`, `ArrayCoordinates1d', and `UniformCoordinates1d`). That would factor out common code that doesn't rely on the `_coords` attribute and use super when appropriate. Then it would be very clear that this code is meant to convert to `StackedCoordinates`. That would probably be the correct way to do this.

Another alternative that works is to implement `RotatedCoordinates._coords` as a property, and then any methods or super calls that need the `_coords` list would work. But it defeats the purpose of parameterizing the coordinates if you make the full coordinates all the time inadvertantly!

File: podpac/core/coordinates/stacked_coordinates.py
Comment (jmilloy) [reply podpac/core/coordinates/stacked_coordinates.py:336]: Yes. You can see in `Node.create_output_array`: the `idims` and `xcoords` are the `dims` and `coords` arguments for `DataArray`...

... huh! Maybe we should call them `xdims`. `idims` is for "index dimensions", but `xdims` would make sense, too. What do you think?

File: podpac/core/coordinates/coordinates1d.py
Comment (mpu-creare) [reply podpac/core/coordinates/coordinates1d.py]: Could you please add a comment to the code to explain this,

Comment (jmilloy) [reply podpac/core/coordinates/coordinates1d.py]: cfa25b6

File: podpac/core/coordinates/rotated_coordinates.py
Comment (mpu-creare) [reply podpac/core/coordinates/rotated_coordinates.py:206]: I get it. This is fine. The old `super` just tripped me up. Please just comment the code to explain that RotatedCoordinates doesn't have coords and you're essentially creating them here, and then accessing through stacked coordinate.s

File: podpac/core/coordinates/stacked_coordinates.py
Comment (mpu-creare) [reply podpac/core/coordinates/stacked_coordinates.py:336]: I leave it up to you. I'm fine with it either way.

**Comment by mpu-creare** [comment]
Let me know when you're happy with this... again... and I'll do the merge. I have a better idea of what's happening on my branch... though I might have to ask you a few questions.

**Comment by jmilloy** [comment]
@mpu-creare This is ready to go again.

**Comment by mpu-creare** [comment]
Great! Thanks. I'm going to finish working through a bug on my branch and then I'll merge.

File: podpac/core/data/datasource.py
Comment (mpu-creare) [inline podpac/core/data/datasource.py:385]: Is this change because the index is now boolean.

Comment (jmilloy) [reply podpac/core/data/datasource.py:385]: This is just for clarity. Maybe I missed something and it needs to be changed back (and explained). The `if` case is when the `index` has more than one element, and the `else` case is when it is has just one element. If it just has one element, what was the `np.max` for?

Comment (mpu-creare) [reply podpac/core/data/datasource.py:385]: Oh... that's because... uh... it's for `max` flavor. 

I blame the diff for now showing me the larger context. 

Thanks.

File: podpac/core/interpolation/interpolators.py
Comment (mpu-creare) [reply podpac/core/interpolation/interpolators.py:371]: As suspected, git had no idea. Fixed.

**Pull request merged**
